### PR TITLE
Implement {Series,DataFrame}GroupBy `fillna` methods

### LINKED ISSUE #8708: Missing Built-in methods for SeriesGroupBy `ffill`
Some of the "built-in" methods for dask SeriesGroupBy objects are missing.

In pandas we could do something like...

```
import numpy as np
import pandas as pd

df = pd.DataFrame({
    'a_cat': list('aabbbaccc'),
    'b_num': [np.nan if i%3!=0 else (i+1) for i in range(9)]
})

df.groupby('a_cat')['b_num'].ffill()
```

But when I try to do it in Dask..

```
import dask.dataframe as dd

ddf = dd.from_pandas(df)
ddf.groupby('a_cat')['b_num'].ffill()
```
I receive an `AttributeError: 'SeriesGroupBy' object has no attribute 'ffill'`

---

I'm hoping this can be implemented along with the other `fillna` methods. Thanks.

### PR DESCRIPTION
**Pull request by pavithraes**
- [x] Closes #8708
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

TODO before merging:
- [x] Add docstring
- [x] Open new issue for implementing value = dict/Series/DataFrame (done: #8922)

### CODE REVIEW

**Review by ian-r-rose — COMMENTED** [review]
Thanks @pavithraes, this is looking great! The implementation is looking good to me, I think we could get some more coverage of the optional args, but otherwise I think this looks close.

I also notice that there are `forwardfill` and `backfill` aliases in pandas, we could think about adding those as well while we're at it.

File: dask/dataframe/groupby.py [hunk 1]
```diff
@@ -1984,6 +1984,19 @@ def rolling(self, window, min_periods=None, center=False, win_type=None, axis=0)
             axis=axis,
         )
 
+    def fillna(self, value=None, method=None, limit=None, axis=None):
```
[the remark this answers is not in the document]
Comment (pavithraes) [reply dask/dataframe/groupby.py:1987]: Thanks, I agree!

File: dask/dataframe/groupby.py [hunk 2]
```diff
@@ -1984,6 +1984,19 @@ def rolling(self, window, min_periods=None, center=False, win_type=None, axis=0)
...
+        fn = lambda g: g.fillna(value=value, method=method, limit=limit, axis=axis)
+        meta = self._meta_nonempty.transform(fn)
+        return self.transform(fn, meta=meta)
+
+    @derived_from(pd.core.groupby.GroupBy)
+    def ffill(self, axis=None, limit=None):
```
[the remark this answers is not in the document]
Comment (pavithraes) [reply dask/dataframe/groupby.py:1993]: Oh yes, thanks for the catch!

**Comment by pavithraes** [comment]
@ian-r-rose Thanks for the review! 

> I also notice that there are forwardfill and backfill aliases in pandas, we could think about adding those as well while we're at it.

Looks like [`pad`](https://github.com/pandas-dev/pandas/blob/v1.4.2/pandas/core/groupby/groupby.py#L2575-L2582) and [`backfill`](https://github.com/pandas-dev/pandas/blob/v1.4.2/pandas/core/groupby/groupby.py#L2611-L2618) has been deprecated in pandas, so I think we don't need to add them?

I also noticed the `transform` implementation doesn't work for `axis=1`, it throws: `ValueError: transform must return a scalar value for each group`. I'm assuming this is because [`transform` applies the function on a column-by-column basis](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.groupby.DataFrameGroupBy.transform.html)?

<details>
<summary>Full traceback</summary>

```python-traceback
ddf.groupby("A").fillna(0, axis=1)
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/core/generic.py in _get_axis_number(cls, axis)
    545         try:
--> 546             return cls._AXIS_TO_AXIS_NUMBER[axis]
    547         except KeyError:

KeyError: 1

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/core/groupby/generic.py in _transform_general(self, func, *args, **kwargs)
   1316             try:
-> 1317                 path, res = self._choose_path(fast_path, slow_path, group)
   1318             except TypeError:

~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/core/groupby/generic.py in _choose_path(self, fast_path, slow_path, group)
   1393         path = slow_path
-> 1394         res = slow_path(group)
   1395 

~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/core/groupby/generic.py in <lambda>(group)
   1386             fast_path = lambda group: func(group, *args, **kwargs)
-> 1387             slow_path = lambda group: group.apply(
   1388                 lambda x: func(x, *args, **kwargs), axis=self.axis

~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/core/frame.py in apply(self, func, axis, raw, result_type, args, **kwargs)
   8739         )
-> 8740         return op.apply()
   8741 

~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/core/apply.py in apply(self)
    687 
--> 688         return self.apply_standard()
    689 

~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/core/apply.py in apply_standard(self)
    811     def apply_standard(self):
--> 812         results, res_index = self.apply_series_generator()
    813 

~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/core/apply.py in apply_series_generator(self)
    827                 # ignore SettingWithCopy here in case the user mutates
--> 828                 results[i] = self.f(v)
    829                 if isinstance(results[i], ABCSeries):

~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/core/groupby/generic.py in <lambda>(x)
   1387             slow_path = lambda group: group.apply(
-> 1388                 lambda x: func(x, *args, **kwargs), axis=self.axis
   1389             )

~/Developer/Dask/dask/dask/dataframe/groupby.py in _fillna_groups(self, groups, **kwargs)
   1987     def _fillna_groups(self, groups, **kwargs):
-> 1988         return groups.fillna(**kwargs)
   1989 

~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/util/_decorators.py in wrapper(*args, **kwargs)
    310                 )
--> 311             return func(*args, **kwargs)
    312 

~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/core/series.py in fillna(self, value, method, axis, inplace, limit, downcast)
   4815     ) -> Series | None:
-> 4816         return super().fillna(
   4817             value=value,

~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/core/generic.py in fillna(self, value, method, axis, inplace, limit, downcast)
   6320             axis = 0
-> 6321         axis = self._get_axis_number(axis)
   6322 

~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/core/generic.py in _get_axis_number(cls, axis)
    547         except KeyError:
--> 548             raise ValueError(f"No axis named {axis} for object type {cls.__name__}")
    549 

ValueError: No axis named 1 for object type Series

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
<ipython-input-4-f9b1a98697a0> in <module>
----> 1 ddf.groupby("A").fillna(0, axis=1)

~/Developer/Dask/dask/dask/dataframe/groupby.py in fillna(self, value, method, limit, axis)
   1993                 "groupby-fillna with value=dict/Series/DataFrame is currently not supported"
   1994             )
-> 1995         meta = self._meta_nonempty.transform(
   1996             self._fillna_groups, value=value, method=method, limit=limit, axis=axis
   1997         )

~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/core/groupby/generic.py in transform(self, func, engine, engine_kwargs, *args, **kwargs)
   1355     @Appender(_transform_template)
   1356     def transform(self, func, *args, engine=None, engine_kwargs=None, **kwargs):
-> 1357         return self._transform(
   1358             func, *args, engine=engine, engine_kwargs=engine_kwargs, **kwargs
   1359         )

~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/core/groupby/groupby.py in _transform(self, func, engine, engine_kwargs, *args, **kwargs)
   1444 
   1445         if not isinstance(func, str):
-> 1446             return self._transform_general(func, *args, **kwargs)
   1447 
   1448         elif func not in base.transform_kernel_allowlist:

~/mambaforge/envs/dask-dev/lib/python3.9/site-packages/pandas/core/groupby/generic.py in _transform_general(self, func, *args, **kwargs)
   1320             except ValueError as err:
   1321                 msg = "transform must return a scalar value for each group"
-> 1322                 raise ValueError(msg) from err
   1323 
   1324             if isinstance(res, Series):

ValueError: transform must return a scalar value for each group
```

</details>

**Comment by ian-r-rose** [comment]
> Looks like [`pad`](https://github.com/pandas-dev/pandas/blob/v1.4.2/pandas/core/groupby/groupby.py#L2575-L2582) and [`backfill`](https://github.com/pandas-dev/pandas/blob/v1.4.2/pandas/core/groupby/groupby.py#L2611-L2618) has been deprecated in pandas, so I think we don't need to add them?

Sounds good to me, good catch.

> I also noticed the transform implementation doesn't work for axis=1

Interesting! Let's take a look in person.

File: dask/dataframe/groupby.py [hunk 3]
```diff
@@ -1984,6 +1984,34 @@ def rolling(self, window, min_periods=None, center=False, win_type=None, axis=0)
             axis=axis,
         )
 
+    def _fillna_groups(self, groups, **kwargs):
```
Comment (ian-r-rose) [inline dask/dataframe/groupby.py:1987]: This function doesn't need to be a member function (note we don't use `self` anywhere), so it's probably cleaner and less likely to run into serialization issues to have it be a top-level function

File: dask/dataframe/tests/test_groupby.py [hunk 4]
```diff
@@ -1091,6 +1091,84 @@ def test_aggregate_dask():
...
+    )
+
+    with pytest.raises(NotImplementedError):
+        ddf.groupby("A").fillna({"A": 0})
+
+    with pytest.raises(NotImplementedError):
+        ddf.groupby("A").fillna(pd.Series(dtype=int))
+
+    with pytest.raises(NotImplementedError):
+        ddf.groupby("A").fillna(pd.DataFrame)
```
Comment (ian-r-rose) [inline dask/dataframe/tests/test_groupby.py:1127]: Now that I think of it, we might be able to get fillna with a dict/pandas object working with relatively little effort.

We could still defer on a Dask Series/DF, however.

File: dask/dataframe/groupby.py (hunk 3 above)
Comment (pavithraes) [reply dask/dataframe/groupby.py:1987]: That makes sense, I'll update it!

File: dask/dataframe/tests/test_groupby.py (hunk 4 above)
Comment (pavithraes) [reply dask/dataframe/tests/test_groupby.py:1127]: In our discussion yesterday, we noticed that value=dict creates a multi-indexed-DF. So, we decided to open a new issue to implement value = pandas dict/series/dataframe. I've added a TODO in the first comment, hence resolving this. :)

File: dask/dataframe/groupby.py [hunk 5]
```diff
@@ -1023,6 +1023,15 @@ def _cumcount_aggregate(a, b, fill_value=None):
     return a.add(b, fill_value=fill_value) + 1
 
 
+def _fillna_groups(groups, axis, **kwargs):
+    if axis == 0:
+        return groups.fillna(axis=axis, **kwargs)
+    else:
+        return groups.drop(columns=groups.columns[0], axis=1).fillna(
```
Comment (pavithraes) [inline dask/dataframe/groupby.py:1030]: @ian-r-rose Is this way of dropping columns alright? I noticed there is a `drop_columns` dataframe-method, would it make sense to use that instead?

https://github.com/dask/dask/blob/1b1a96b4ddf31d0879e718be463a731af7f1d086/dask/dataframe/methods.py#L379-L382

Comment (pavithraes) [reply dask/dataframe/groupby.py:1030]: And, I'm also not sure about bundling drop-column logic in the top-level `def _fillna_groups()`. is this OK, or would it be better to do this in `fillna()` (line 2014)?

Comment (pavithraes) [reply dask/dataframe/groupby.py:1030]: Just to note: I'm yet to add doctrings and comments :)

File: dask/dataframe/groupby.py [hunk 6]
```diff
@@ -1984,6 +1993,45 @@ def rolling(self, window, min_periods=None, center=False, win_type=None, axis=0)
             axis=axis,
         )
 
+    def fillna(self, value=None, method=None, limit=None, axis=None):
+        if not np.isscalar(value) and value is not None:
+            raise NotImplementedError(
+                "groupby-fillna with value=dict/Series/DataFrame is currently not supported"
+            )
+        axis = DataFrame._validate_axis(axis)
+        if value is not None or axis == 0:
```
Comment (pavithraes) [inline dask/dataframe/groupby.py:2002]: If we use `value` with `fillna` the `axis` doesn't really matter because the output remains the same.

But, that said, I feel if a user specifies `axis=1` they probably assume Dask to be working row-wise internally, so maybe we should use the `apply` implementation whenever `axis=1` irrespective of `value`?

Comment (ian-r-rose) [reply dask/dataframe/groupby.py:2002]: I think your choice here is fine, I don't want to belabor the point of `apply` here, it's more of an implementation detail, and shouldn't really affect the user much.

File: dask/dataframe/groupby.py [hunk 7]
```diff
@@ -1023,6 +1023,15 @@ def _cumcount_aggregate(a, b, fill_value=None):
     return a.add(b, fill_value=fill_value) + 1
 
 
+def _fillna_groups(groups, axis, **kwargs):
```
Comment (ian-r-rose) [inline dask/dataframe/groupby.py:1026]: Nit: this is applying to a single group, so I think `group` would be a more appropriate name

Comment (ian-r-rose) [reply dask/dataframe/groupby.py:1026]: Rather than a `**kwargs`, I'd just include the rest of the args explicitly

File: dask/dataframe/groupby.py (hunk 5 above)
Comment (ian-r-rose) [reply dask/dataframe/groupby.py:1030]: I think using `df.drop` directly is okay. It looks like the `drop_columns` utility is primarily used for `map_partitions`, which doesn't really apply here.

However, I think always dropping the zeroth column doesn't quite do what we want:
1. The column we want to drop might not be the first one.
2. The user might be grouping by multiple columns

What if instead we also included a `by` argument in `_fillna_group` for the column(s) by which we are grouping, the dataframe can pass in `self.by`, and then we can drop them?

**Review by ian-r-rose — COMMENTED** [review]
Thanks @pavithraes, this is looking close.

I did some experimenting with grouping by multiple columns, and ran into some troubles with `transform` not liking empty groups very much, but `apply` did just fine. I wonder if we should just simplify things by using `apply` everywhere

File: dask/dataframe/groupby.py (hunk 5 above)
Comment (pavithraes) [reply dask/dataframe/groupby.py:1030]: > What if instead we also included a by argument in _fillna_group for the column(s) by which we are grouping, the dataframe can pass in self.by, and then we can drop them?

Yes! Thanks for this suggestion!

File: dask/dataframe/groupby.py (hunk 7 above)
Comment (pavithraes) [reply dask/dataframe/groupby.py:1026]: Sounds good to me, shall do!

**Comment by pavithraes** [comment]
> I did some experimenting with grouping by multiple columns, and ran into some troubles with transform not liking empty groups very much, but apply did just fine. I wonder if we should just simplify things by using apply everywhere

Thanks for checking! I agree that we can use apply throughout. I'll also include a test for empty groups and grouping by multiple columns. :)

**Comment by pavithraes** [comment]
> I did some experimenting with grouping by multiple columns, and ran into some troubles with transform not liking empty groups very much, but apply did just fine. I wonder if we should just simplify things by using apply everywhere

Thanks for checking! I agree that we can use apply throughout. I'll also include a test for grouping by multiple columns. :)

File: dask/dataframe/groupby.py [hunk 8]
```diff
@@ -1984,6 +1990,65 @@ def rolling(self, window, min_periods=None, center=False, win_type=None, axis=0)
             axis=axis,
         )
 
+    def fillna(self, value=None, method=None, limit=None, axis=None):
+        """Fill NA/NaN values using the specified method.
```
Comment (ian-r-rose) [inline dask/dataframe/groupby.py:1994]: Does the `@derived_from` decorator not work here?

File: dask/dataframe/groupby.py [hunk 9]
```diff
@@ -1023,6 +1023,12 @@ def _cumcount_aggregate(a, b, fill_value=None):
     return a.add(b, fill_value=fill_value) + 1
 
 
+def _fillna_group(group, by, value, method, limit, fillna_axis):
+    return group.drop(columns=by).fillna(
```
Comment (ian-r-rose) [inline dask/dataframe/groupby.py:1027]: I think a comment as to why we are dropping columns would be useful, since it's not super obvious to the reader (namely, that `apply` keeps the grouped-by columns around)

**Review by ian-r-rose — APPROVED** [review]
Thanks @pavithraes, this looks great to me! I have a couple of minor docs suggestions, which I don't think are required

**Comment by ian-r-rose** [comment]
@pavithraes [This change](https://github.com/pandas-dev/pandas/pull/34998/files) in upstream pandas might also affect what you are doing here.

File: dask/dataframe/groupby.py (hunk 8 above)
Comment (pavithraes) [reply dask/dataframe/groupby.py:1994]: It does, but since we don't support all the `value` types and all keyword arguments, I decided to create a custom docstring -- which copies the pandas' version, just with some edits.

File: dask/dataframe/groupby.py (hunk 9 above)
Comment (pavithraes) [reply dask/dataframe/groupby.py:1027]: Sounds good, added!

**Comment by pavithraes** [comment]
> [This change](https://github.com/pandas-dev/pandas/pull/34998/files) in upstream pandas might also affect what you are doing here.

Thanks for the note! I just triggered an upstream build to check and it does seem to be affected. :/

**Comment by pavithraes** [comment]
@jrbourbeau Thanks for helping review this!

To solve the upstream build issues, we need to add `group_keys=False` (it's False because `fillna` disregards the columns we grouped-by in the output) to the `apply` calls in this PR.

I think we can't add that right now because [Dask's groupby-apply](https://docs.dask.org/en/stable/generated/dask.dataframe.groupby.DataFrameGroupBy.apply.html#dask.dataframe.groupby.DataFrameGroupBy.apply) doesn't support the `group_keys` parameter yet, so we may need to resolve this with the other failing pandas-upstream-tests in #8875?

**Review by ian-r-rose — COMMENTED** [review]
I rebased your PR on top of #8961 to see if it still works with the `group_keys` changes, and it does :tada: !

The bad news is: with the `group_keys` changes I think there is one more thing we need to cover. I tried parameterizing your `test_fillna` over `group_keys=[True, False, None]`. Everything works on pandas `1.4.x`, but on `1.5.x`, `group_keys=True` presents a problem: in that case, the `apply` returns with extra multiindex columns for your grouped-by items.

I think the fix should be fairly straightforward, however. For the case where `self.group_keys` is `True` in the `dd.GroupBy` object, we can do a final `map_partitions` to `df.droplevel(by)`. That is to say, we can easily drop the extra index columns at the end when `group_keys` is `True`. It's annoying, but doable

**Comment by pavithraes** [comment]
@ian-r-rose Thanks for the notes! I've pushed the `map_partitions` fix that you suggested, and it seems to work locally with `pandas 1.4.2` and `1.5.0`!

**Comment by pavithraes** [comment]
I'm not sure why the macos-python-3.8 test is failing with: https://github.com/dask/dask/runs/6209266570?check_suite_focus=true#step:6:21560 :/

File: dask/dataframe/groupby.py [hunk 10]
```diff
@@ -2008,6 +2016,71 @@ def rolling(self, window, min_periods=None, center=False, win_type=None, axis=0)
...
+            by=self.by,
+            value=value,
+            method=method,
+            limit=limit,
+            fillna_axis=axis,
+            meta=meta,
+        )
+
+        if PANDAS_GT_150 and self.group_keys:
+            return result.map_partitions(lambda df: df.droplevel(self.by))
```
Comment (ian-r-rose) [inline dask/dataframe/groupby.py:2072]: This is great! One minor suggestion: there is a `MethodCache` (goes by `M` here) which might be slightly faster than this lambda. I doubt it makes a huge difference either way, though.

```suggestion
            return result.map_partitions(M.droplevel, self.by)
```

**Review by ian-r-rose — APPROVED** [review]
Sorry to be slow in reviewing @pavithraes, and thanks for your persistence. This looks great! I have one minor comment that I would consider optional. But from my perspective, this is ready to go into our next release!

File: dask/dataframe/groupby.py (hunk 10 above)
Comment (pavithraes) [reply dask/dataframe/groupby.py]: Thank you, this is much nicer!

**Review by ian-r-rose — APPROVED** [review]
Thanks @pavithraes!

**Comment by pavithraes** [comment]
@jrbourbeau I think this is ready to merge! Could you please help take a quick look?

**Review by jrbourbeau — APPROVED** [review]
Thanks @pavithraes and @ian-r-rose! There's a merge conflict that's popped up (apologies for the delayed response). @pavithraes would you mind fixing that conflict? Otherwise, this looks good to go

**Comment by pavithraes** [comment]
@jrbourbeau Thanks for taking a look! I've fixed the conflict. :)

**Comment by ian-r-rose** [comment]
Thank you @pavithraes!
